File: js-assembler/src/instructions/jumps.js
/**
 * Jump instruction handlers (JP, JR, DJNZ)
 */

import { parseImmediate } from '../utils/parseImmediate.js';

export function handleJumpInstruction(mnemonic, operands, context) {
  const result = [];
  const { labels, constants, currentAddress } = context;

  // JP instruction (with conditional support)
  if (mnemonic === "jp") {
    const parts = operands.split(",").map(p => p.trim());
    if (parts.length === 2) {
      // Conditional jump
      const conditionCodes = {
        "z": 0xca, "nz": 0xc2, "c": 0xda, "nc": 0xd2,
        "pe": 0xea, "po": 0xe2, "p": 0xf2, "m": 0xfa
      };
      if (conditionCodes[parts[0]]) {
        const address = labels[parts[1]] !== undefined
          ? labels[parts[1]]
          : parseImmediate(parts[1], constants);
        result.push(conditionCodes[parts[0]]);
        result.push(address & 0xff);
        result.push((address >> 8) & 0xff);
        return new Uint8Array(result);
      }
    } else {
      // Unconditional jump
      const address =
        labels[operands] !== undefined
          ? labels[operands]
          : parseImmediate(operands, constants);
      result.push(0xc3); // JP nn
      result.push(address & 0xff);
      result.push((address >> 8) & 0xff);
      return new Uint8Array(result);
    }
  }

  // JR instruction (relative jump with conditional support)
  if (mnemonic === "jr") {
    const parts = operands.split(",").map(p => p.trim());
    let opcode = 0x18; // Default: unconditional JR
    let target;
    
    if (parts.length === 2) {
      // Conditional JR
      const conditions = {
        "z": 0x28, "nz": 0x20, "c": 0x38, "nc": 0x30
      };
      if (conditions[parts[0]]) {
        opcode = conditions[parts[0]];
        target = parts[1];
      } else {
        throw new Error(`Invalid JR condition: ${parts[0]}`);
      }
    } else {
      // Unconditional JR
      target = operands;
    }
    
    // Calculate relative offset
    let offset;
    if (labels[target] !== undefined) {
      offset = labels[target] - (currentAddress + 2);
    } else {
      // Try to parse as immediate offset
      offset = parseImmediate(target, constants);
    }
    
    if (offset < -128 || offset > 127) {
      throw new Error(`JR offset too large: ${offset}`);
    }
    
    result.push(opcode);
    result.push(offset & 0xff);
    return new Uint8Array(result);
  }
  
  // DJNZ instruction (decrement B and jump if not zero)
  if (mnemonic === "djnz") {
    let offset;
    if (labels[operands] !== undefined) {
      offset = labels[operands] - (currentAddress + 2);
    } else {
      offset = parseImmediate(operands, constants);
    }
    
    if (offset < -128 || offset > 127) {
      throw new Error(`DJNZ offset too large: ${offset}`);
    }
    
    result.push(0x10); // DJNZ e
    result.push(offset & 0xff);
    return new Uint8Array(result);
  }

  return null;
}

File: rust-assembler/src/instructions/jumps.rs
use crate::utils::immediate::parse_immediate;
use anyhow::Result;
use std::collections::HashMap;

pub fn handle_jump_instruction(
    mnemonic: &str,
    operands: Option<&str>,
    labels: &HashMap<String, u16>,
    constants: &HashMap<String, u16>,
    current_address: u16,
) -> Result<Option<Vec<u8>>> {
    let mut result = Vec::new();

    match mnemonic {
        "jp" => {
            if let Some(ops) = operands {
                // Handle conditional jumps
                let (condition, target) = if ops.contains(',') {
                    let parts: Vec<&str> = ops.split(',').map(|s| s.trim()).collect();
                    (Some(parts[0]), parts[1])
                } else {
                    (None, ops.trim())
                };

                let address = if let Some(&label_addr) = labels.get(target) {
                    label_addr
                } else {
                    parse_immediate(target, constants)?
                };

                // Encode conditional jump
                match condition {
                    Some("nz") => result.push(0xc2),
                    Some("z") => result.push(0xca),
                    Some("nc") => result.push(0xd2),
                    Some("c") => result.push(0xda),
                    Some("po") => result.push(0xe2),
                    Some("pe") => result.push(0xea),
                    Some("p") => result.push(0xf2),
                    Some("m") => result.push(0xfa),
                    None => result.push(0xc3), // Unconditional JP
                    _ => return Ok(None),
                }

                result.push((address & 0xff) as u8);
                result.push(((address >> 8) & 0xff) as u8);
                return Ok(Some(result));
            }
        },
        "jr" => {
            if let Some(ops) = operands {
                // Handle conditional relative jumps
                let (condition, target) = if ops.contains(',') {
                    let parts: Vec<&str> = ops.split(',').map(|s| s.trim()).collect();
                    (Some(parts[0]), parts[1])
                } else {
                    (None, ops.trim())
                };

                let target_addr = if let Some(&label_addr) = labels.get(target) {
                    label_addr
                } else {
                    parse_immediate(target, constants)?
                };

                // Calculate relative offset
                // JR instruction is 2 bytes, offset is from the next instruction
                let offset = (target_addr as i32) - (current_address as i32 + 2);
                if offset < -128 || offset > 127 {
                    return Err(anyhow::anyhow!("JR target out of range: offset {}", offset));
                }

                // Encode conditional relative jump
                match condition {
                    Some("nz") => result.push(0x20),
                    Some("z") => result.push(0x28),
                    Some("nc") => result.push(0x30),
                    Some("c") => result.push(0x38),
                    None => result.push(0x18), // Unconditional JR
                    _ => return Ok(None),
                }

                result.push(offset as u8);
                return Ok(Some(result));
            }
        },
        "djnz" => {
            if let Some(target) = operands {
                let target_addr = if let Some(&label_addr) = labels.get(target.trim()) {
                    label_addr
                } else {
                    parse_immediate(target.trim(), constants)?
                };

                // Calculate relative offset
                let offset = (target_addr as i32) - (current_address as i32 + 2);
                if offset < -128 || offset > 127 {
                    return Err(anyhow::anyhow!(
                        "DJNZ target out of range: offset {}",
                        offset
                    ));
                }

                result.push(0x10); // DJNZ
                result.push(offset as u8);
                return Ok(Some(result));
            }
        },
        _ => {},
    }

    Ok(None)
}
